### PR TITLE
Test routine: add Makefile; CI: consolidate ci.yaml files

### LINKED ISSUE #1052: Refactor test routine
## Checklist

- [x] I am using the current [`main`](https://github.com/pypsa-meets-earth/pypsa-earth/tree/main) branch or the latest [release](https://github.com/pypsa-meets-earth/pypsa-earth/releases). Please indicate.
- [x] I am running on an up-to-date [`pypsa-earth` environment](https://github.com/pypsa-meets-earth/pypsa-earth/blob/main/envs/environment.yaml). Update via `conda env update -f envs/environment.yaml`.

## Problem Statement

Currently the test routine uses nested snakemake calls where snakemake is called via a bash command in the Snakefile. This architecture omits error streams and snakemake runs that fail are not reported as such, see https://github.com/pypsa-meets-earth/pypsa-earth/actions/runs/9667460555/job/26669336528#step:9:1739.

## Suggestion

I would propose to execute `snakemake` in the GH workflow and optionally keep the commands in a common file for easy local execution, like https://github.com/PyPSA/pypsa-eur/blob/master/test.sh

I would as well propose to collect all workflow CI's in one yaml that uses a `matrix` entry for setting the operation systems.

### PR DESCRIPTION
**Pull request by FabianHofmann**
Closes #1052 

## Changes proposed in this Pull Request

This PR proposes four changes:

 - fix a bug in `add_extra_components`. CSP generators are not necessarily attached to every bus (this is probably related to a "missing" substation or shape assigment). This lead to an error in the CI before (as mentioned in #1052) in the alignment of `buses_i` being the AC-buses and the csp buses.
 - use a Makefile to setup and run tests. With `make test` all the snakemake tests can now be run locally. Why is that better? The workflow used a recursive snakemake call. This lead to errors not being reported back as errors. With the Makefile, errors are strightly reported back and the test command are explicit and easily accessible.
 - the `run_scenario` rule now streams errors which it did not do before. At some point, we might also think using something similar as for the tests to improve visibility of the command used. 
 -  Consolidate CI yaml. Use only one CI yaml instead of three. Use micromamba with it built in cashing functionality. 

In general, the workflow would profit from using `--configfile=<additional.config.yaml>` instead of the diff_config approach. Was there a specific reason to stick to the manual insertion of config diffs? Perhaps is that historically grown? Note that `--configfile` overrides existing and adds additonal keys to the config given in the Snakefile. So it does essentially the same.     

The CI will likely fail now, I need probably some more iterations. 

## Checklist

- [x] I consent to the release of this PR's code under the AGPLv3 license and non-code contributions under CC0-1.0 and CC-BY-4.0.
- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `doc/requirements.txt`.
- [x] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [x] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by pz-max** [comment]
I can comment here a bit:
- @FabianHofmann , we split the CI to test everything in Linux (faster) & only a subset in other OS. Another reason was the need for a badge for each OS. This can be useful e.g. when you run on Linux where all tests are passing and you don't care about Windows too much.
- love the `make` solution 

Regarding this bit:
> In general, the workflow would profit from using --configfile=<additional.config.yaml> instead of the diff_config approach. Was there a specific reason to stick to the manual insertion of config diffs? Perhaps is that historically grown? Note that --configfile overrides existing and adds additonal keys to the config given in the Snakefile. So it does essentially the same

I worked on the config diff stuff. Snakemake uses the exact same config merging code as we do (realised that later). I think @davide-f possibly added a few other parts. We implemented it like this to be independent of Snakemake and more flexible for this functionality (e.g. we are using ruaml (as Yaml container), or smt like that, which also keeps comments after the config is merged). Not sure what other benefits/ disadvantages exist. @davide-f might have more thoughts. 

Feel free to decide what is the better solution for now :+1:

File: Snakefile
Comment (davide-f) [inline Snakefile]: Are we sure about this?
I'm wondering that it may be good to preserve.
What do you think?
I like though specifying the three configfile: as proposed but probably within the if?

**Review by davide-f — COMMENTED** [review]
Many thanks @FabianHofmann !!! Much needed!
Added a comment, happy to talk tomorrow if you plan to participate to the meeting [I understood so or I'm wrong?]

However the CI is not working; are you sure the configfiles get merged well?

File: Snakefile
Comment (FabianHofmann) [reply Snakefile]: Hey @davide-f, thanks for your reply! we are getting there :) This is indeed one change that I am not sure about. The reason why I put this in is that I think the workflow should also be tested against the `config.default.yaml` settings. Before, the CI was "only" testing the tutorial config as a basis. This leads to the artefact that one always has to secure manually that the default config is in line with the tutorial config. The proposed approach is the same as we have it in pypsa-eur and uses 

- config.default.yaml as level-zero defaults
- add configs/bundle_config.yaml which extends the default (no overrides)
- use incremental config changes from the user via `config.yaml`. Here the user can only set the keys which should be different from the `config.default.yaml`. In a python function, that would translate to setting a keyword-argument which already has a default to something else. So the `config.yaml` could be empty, and the workflow would still run.   

The structure ensure that the user always has a stable set of default configs. Whenever there is an update to the default, the user does not need to copy it to `config.yaml` and insert their custom settings manually, it happens automatically.  

However, it leads to one or two more needed fixes in the workflow. For example, the CI is breaking now as the solver_options are not fully compatible. Snakemake is currently merging the solver options from the default solver with the ones from the tutorial solver. 

Two options:

1. Go full in -> make an adoption of the solver_options as we have it in pypsa-eur (already happened and will be happening anyway through #796 )
2. Revert this setting and introduce it at a later point (which I would strongly encourage to do)

what to you think?

**Comment by FabianHofmann** [comment]
@pz-max @davide-f this is ready for being merged, objective values are reproduced. A merge today would be cool, so I can look into #796 soon

File: Snakefile
Comment (ekatef) [reply Snakefile]: Hello @FabianHofmann! Sounds like a well-elaborated and quite an elegant solution.

However, the idea to automate introducing changes into `config.yaml` sounds quite alarming for me. If feels like we are making the user interface unstable (because `config.yaml` is actually the major UI of the model)  which is a famous architectural anti-pattern and must be avoided.

According to the user experience we have observed, the only use case when such automation may be done in a safe way, is when `config.yam` doesn't exist. Exactly that is a case addressed by the initial configuration. So, I agree with @davide-f that the changes would be better placed under `if` condition , as it was before. What do you think?

File: doc/release_notes.rst
Comment (ekatef) [inline doc/release_notes.rst:52]: Actually, the workflow still has functionality which allows for incrementally changes in the configuration files. However, this functionality is implemented using a dedicated function which places it among advanced stuff which you are using only if you well understand what you are doing. That allows to avoid side effects for the users interested in basic modelling functionality. 

I recognise that there is an approach close to what is used in PyPSA-Eur, but according to out observations for the requests in `Support` channel, that is not a concept which is easy for use.

File: envs/environment.yaml
Comment (ekatef) [inline envs/environment.yaml:80]: As we discussed, that would lead to workflow crashes for Windows users, unfortunately.

File: config.default.yaml
Comment (ekatef) [inline config.default.yaml:435]: @FabianHofmann has it been a specific reason for adding these updates to the PR? 

The linopy-integration PR #796 introduces the same changes, so we'll have them in the workflow in any case. Do you see any advantages for testing from having these changes earlier? Just trying to understand the logic :)

**Review by ekatef — COMMENTED** [review]
Hello @FabianHofmann, thanks a lot for the great contribution!

Have added a couple of questions related to some details. Could you please look into them? My general feeling is that it would be great to avoid breaking changes, especially those which relate to the behaviour of user interfaces.

File: Snakefile
Comment (FabianHofmann) [reply Snakefile]: hey @ekatef! Thanks for the review :) Perhaps I did not explain myself too well. So, there will be now breaking changes to the workflow. As well, the workflow will not automatically change the config.yaml. It is just the order and config updates that changes. Instead of one config files, the workflow now uses multiples to ensure a full set of defaults and allow the user to make use of config diffs.  The user can now use 

* the config.yaml  as a pure "diff" config to the defaults which only contains effective changes set by the user
* the `snakemake --configfile <myconfig.yaml>`  options

See for example the new [config.tutorial](https://github.com/open-energy-transition/pypsa-earth/blob/refactor-test-and-ci/config.tutorial.yaml). It only contains effective diffs to the default. This makes it much easier to maintain both the default and tutorial config and allows to actually see what are the effective changes to defaults coming from the config.tutorial.   

Old `config.yaml` file will still work. In case there is no `config.yaml` I added a function `copy_default_files()` which sets an empty config.yaml in place as a placeholder (see some lines below). The only exception is the settings of the solver options which anyway comes with the linopy integration which are necesary to allow for config updates.

If you want we can also chat about it later, if that suits :)

File: doc/release_notes.rst
Comment (FabianHofmann) [reply doc/release_notes.rst:52]: I hope I understand the point correctly. Yes, the internal scenario config update routine (sorry for the long term) I did not touch. That is still the same as before. I only touched the "base" config to be the "config.default" but the scenario building is compatible with it. I mainly focused on the test routine and the allowing for incremental config changes (they are not necessarily required, old config behaviour still works)

File: envs/environment.yaml
Comment (FabianHofmann) [reply envs/environment.yaml:80]: yes, this was intended (but forgot to mention it somewhere), I referred to davides comment that we temporarily disable the windows stability until the linopy integration where we can kick out ipopt

File: config.default.yaml
Comment (FabianHofmann) [reply config.default.yaml:435]: Yes, these are needed to enable the incremental config changes (see above). But this will be easy to merge with the linopy PR

**Comment by finozzifa** [comment]
Indeed a great PR @FabianHofmann! thanks a lot

**Comment by FabianHofmann** [comment]
@davide-f @ekatef the PR is now ready. As discussed the `config.default.yaml` is ignored by now, and if no `config.yaml` is existent, the `config.tutorial.yaml` is copy to `config.yaml`

**Comment by ekatef** [comment]
> @davide-f @ekatef the PR is now ready. As discussed the `config.default.yaml` is ignored by now, and if no `config.yaml` is existent, the `config.tutorial.yaml` is copy to `config.yaml`

Hey @FabianHofmann, thank you so much for the discussion and the adjustments! I love the solution with commenting-out the automated update using `config.default.yaml`. Absolutely agree that improvements are needed on management of the configuration files, and having this like in Snakemake should be an effective reminder! Could we probably add an issue to track the ideas and the progress on that, also accounting for the modularity idea @davide-f has suggested?

**Comment by FabianHofmann** [comment]
@ekatef that is a good idea! thank you for your comment. I will raise an issue

**Review by davide-f — APPROVED** [review]
Great @FabianHofmann :D

**Comment by davide-f** [comment]
Thanks @FabianHofmann !
Agree, let's keep it like this for now and edit after the merge :)

**Comment by FabianHofmann** [comment]
mmh, annoying, as soon as passing the `--configfile` via snakemake, the condition 

```
if "config" not in globals() or not config:  # skip when used as sub-workflow
```

is false, so it will not load the `config.yaml`. It reverted the latest changes and have a look on monday, whether the if clause is really necessary for pypsa-earth-sec, or how we can adjust it.

**Comment by davide-f** [comment]
> mmh, annoying, as soon as passing the `--configfile` via snakemake, the condition
> 
> ```
> if "config" not in globals() or not config:  # skip when used as sub-workflow
> ```
> 
> is false, so it will not load the `config.yaml`. It reverted the latest changes and have a look on monday, whether the if clause is really necessary for pypsa-earth-sec, or how we can adjust it.

Unfortunately it is necessary, I had to add that if clause exactly because of that unfortunately.
Worth pointing @hazemakhalek to agree on a solution. If we merge this, we cannot update -sec to latest main here
The merge will solve this issue...

**Comment by davide-f** [comment]
Hello @FabianHofmann , just to confirm, we can close this PR as it is merged already into the merge right?

**Comment by FabianHofmann** [comment]
Correct!

**Pull request closed without merging**
